### PR TITLE
[QPE-451] label rotation

### PR DESCRIPTION
**Pull request by ahmed-bazzara**
JIRA ticket
https://tretton37.atlassian.net/secure/RapidBoard.jspa?rapidView=33&projectKey=QPE&modal=detail&selectedIssue=QPE-451

### CODE REVIEW

File: src/ldw-barsPlus.js [hunk 1]
```diff
@@ -691,22 +691,30 @@ export default {
...
         .enter()
         .append("text")
         .attr("class", "ldwtxt")
         .style("opacity", "0")
-        .each(function (d) {
+        .each(function (d, i) {
+
           var txp = g.barText(d);
+          var bar = g.bars[0].find(e=>{
+            return e.__data__.dim1 == d.dim1;
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: Switch to strict equality check or comment on why it can't be

File: src/ldw-barsPlus.js [hunk 2]
```diff
@@ -879,8 +887,15 @@ export default {
...
 */
   barText: function (d, total) {
-    var tx, txt, origX, bHeight, textX, bb, textY, textLength, ts;
+
+    var tx, txt, origX, textX, bb, textY, textLength, ts;
     var g = this;
+    var rotation = false;
+    var isElip = false;
+    let bHeight;
+    // var bar = g.bars[0].find(e=>{
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: clean up

File: src/ldw-barsPlus.js [hunk 3]
```diff
@@ -942,21 +962,58 @@ export default {
...
+              txt = '';
+            }
+          }
+        }
+      }
+      if (!g.rotateLabel && g.textDots){
         textLength = g.tref.node().getComputedTextLength();
         txt = g.tref.text();
-        while (textLength > textX - 2 * innerBarPadH && txt.length > 0) {
+        while (textLength > textX - 2 * innerBarPadH && txt.length > 0) {// bHeight -=60;
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: clean up

File: src/ldw-barsPlus.js [hunk 4]
```diff
@@ -691,22 +691,30 @@ export default {
...
     if (~"BA".indexOf(g.showTexts)) {
       // Create text inside bars
+
       g.texts
         .enter()
         .append("text")
         .attr("class", "ldwtxt")
         .style("opacity", "0")
-        .each(function (d) {
+        .each(function (d, i) {
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: rename d and i

File: src/ldw-barsPlus.js [hunk 5]
```diff
@@ -691,22 +691,30 @@ export default {
...
+
           var txp = g.barText(d);
+          var bar = g.bars[0].find(e=>{
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: rename e

File: src/ldw-barsPlus.js [hunk 6]
```diff
@@ -942,21 +962,58 @@ export default {
...
+          let textOnCharecterPixleRatio = textLength / numberOfTextCharacters;
+
+          if ( remainingSpaceForText < 0) {
+            remainingSpaceForText = 0;
+          }
+          if (bHeight - extraPadding > textLength){
+            txt = g.tref.text();
+          }
+          else{
+            let textThatShouldbeEliip = textLength - remainingSpaceForText ;
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: typo

File: src/ldw-barsPlus.js [hunk 7]
```diff
@@ -879,8 +887,15 @@ export default {
  * Get bar text information: x, y and text
 */
   barText: function (d, total) {
-    var tx, txt, origX, bHeight, textX, bb, textY, textLength, ts;
+
+    var tx, txt, origX, textX, bb, textY, textLength, ts;
     var g = this;
+    var rotation = false;
+    var isElip = false;
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: typo

File: src/ldw-barsPlus.js [hunk 8]
```diff
@@ -942,21 +962,58 @@ export default {
...
+          if (bHeight - extraPadding > textLength){
+            txt = g.tref.text();
+          }
+          else{
+            let textThatShouldbeEliip = textLength - remainingSpaceForText ;
+            let numberOfChartoBeEllip = Math.ceil(textThatShouldbeEliip / textOnCharecterPixleRatio) +20;
+
+            isElip = true;
+            txt = txt.slice(0, -numberOfChartoBeEllip);
+            txt +='\u2026';
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: put '\u2026' in a variable so that it can be named

File: src/ldw-barsPlus.js [hunk 9]
```diff
@@ -942,21 +962,58 @@ export default {
...
+
+      if(g.rotateLabel && barWidth > 25) {
+        if (g.textDots) {
+          textLength = g.tref.node().getComputedTextLength();
+          txt = g.tref.text();
+          const ellipsisLength = 25;
+          const extraPadding= 15;
+          let remainingSpaceForText = bHeight - ellipsisLength -extraPadding ;
+          let numberOfTextCharacters = txt.length;
+          let textOnCharecterPixleRatio = textLength / numberOfTextCharacters;
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: typo

**Review by kristofferlind — CHANGES_REQUESTED** [review]

File: src/ldw-barsPlus.js [hunk 10]
```diff
@@ -697,15 +697,15 @@ export default {
         .append("text")
         .attr("class", "ldwtxt")
         .style("opacity", "0")
-        .each(function (d, i) {
+        .each(function (dataObject) {
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js:700]: thats really generic, its doing foreach on g.texts, is that textNodes?

File: src/ldw-barsPlus.js [hunk 11]
```diff
@@ -971,7 +968,7 @@ export default {
           const extraPadding= 15;
           let remainingSpaceForText = bHeight - ellipsisLength -extraPadding ;
           let numberOfTextCharacters = txt.length;
-          let textOnCharecterPixleRatio = textLength / numberOfTextCharacters;
+          let textOnCharacterPixleRatio = textLength / numberOfTextCharacters;
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: another typo in there Pixle -> Pixel

File: src/ldw-barsPlus.js [hunk 12]
```diff
@@ -980,13 +977,13 @@ export default {
...
+            let textThatShouldbeEllip = textLength - remainingSpaceForText ;
+            let numberOfChartoBeEllip = Math.ceil(textThatShouldbeEllip / textOnCharacterPixleRatio) +20;
 
-            isElip = true;
+            isEllip = true;
             txt = txt.slice(0, -numberOfChartoBeEllip);
-            txt +='\u2026';
-            if (txt == '\u2026'){
+            txt +=ellipsis;
+            if (txt == ellipsis){
```
Comment (kristofferlind) [inline src/ldw-barsPlus.js]: can this be strict equal?

**Review by kristofferlind — CHANGES_REQUESTED** [review]

File: src/ldw-barsPlus.js (hunk 11 above)
Comment (ahmed-bazzara) [reply src/ldw-barsPlus.js]: i love typos :P

File: src/ldw-barsPlus.js (hunk 10 above)
Comment (ahmed-bazzara) [reply src/ldw-barsPlus.js:700]: Nope , it is a data object :/
a smaple:
`{
dim1: "Philippines"
dim2: "Philippines"
offset: 0
qElemNumber: 62
qNum: 6139236.168160478
qText: "6139236.2"
qTextPct: ""
}`

**Review by kristofferlind — APPROVED** [review]

**Comment by johnlunde** [comment]
Test failed. See comment in JIRA ticket

**Pull request merged**